crates/bin/pd/src/migrate/testnet74.rs

```diff
@@ -37,11 +37,40 @@ async fn fix_arb_execution_outputs(delta: &mut StateDelta<Snapshot>) -> anyhow::
     Ok(())
 }
 
+/// Update base liquidity index values to be proto-encoded. Previously they were stored as big-endian
+/// encoded amounts, but in https://github.com/penumbra-zone/penumbra/pull/4188 they were changed
+/// to be proto-encoded.
+///
+/// This will rewrite all values under the `dex/ab/` prefix to be proto-encoded.
+async fn rewrite_base_liquidity_indices(delta: &mut StateDelta<Snapshot>) -> anyhow::Result<()> {
+    let prefix_key = "dex/ab/".as_bytes();
+    tracing::trace!(prefix_key = ?EscapedByteSlice(&prefix_key), "updating base liquidity index values");
+    let mut liquidity_stream = delta.nonverifiable_prefix_raw(&prefix_key).boxed();
+
+    while let Some(r) = liquidity_stream.next().await {
+        let (key, raw_amount): (Vec<u8>, Vec<u8>) = r?;
+        tracing::info!(?key, raw_amount = ?EscapedByteSlice(&raw_amount), "migrating base liquidity index entry");
+
+        let amount = Amount::from_be_bytes(raw_amount.as_slice().try_into()?);
+
+        // Store the correctly formatted new value:
+        delta.nonverifiable_put(key.clone(), amount);
+        tracing::info!(
+            key = ?EscapedByteSlice(&key),
+            raw_amount = ?EscapedByteSlice(&raw_amount),
+            ?amount,
+            "updated base liquidity index"
+        );
+    }
+
+    Ok(())
+}
```

Comment:
Is this migration fn idempotent? We've been telling folks our migrations are idempotent, but not sure that's the case for this one. Zooming out, the changes in #4339 definitely aren't idempotent, so not blocking on this.

Reply:
Good catch. No, it is not idempotent. I wasn't aware that was a requirement for migrations; I will follow-up

Reply:
How is this not idempotent? if you run it multiple times you will get the same result, no? why not?

Reply:
is it because the encoding changes? that's completely fine imo

Reply:
> How is this not idempotent? if you run it multiple times you will get the same result, no? why not? 

Looking more closely, I am not sure at the moment. If we attempt to decode protobuf encoded Amount as a BE-encoded Amount it could fail silently and the Amount would change. However, this requires the protobuf byte encoding to be the same length as the big-endian encoding.

Reply:
Honestly not completely sure this is important but we can try to first decode it as a proto amount and return early if it works, and proceed if it correctly returns an error. This way it is resistant to accidental double runs

Reply:
There are some edge cases:

```rust
        let a: Amount = 4951760157141521099596496895u128.into();

        println!("amount 1: {:?}", a);

        let proto_a_bytes: Vec<u8> = penumbra_proto::core::num::v1::Amount::from(a).encode_to_vec();

        let amount = Amount::from_be_bytes(proto_a_bytes.as_slice().try_into()?);

        println!("amount 2: {:?}", amount);
```

```console
amount 1: 4951760157141521099596496895

amount 2: 11963051962064242856133983240072462207
```

so it could _theoretically_ still munge some values and be non-idempotent even if we took that approach if everything aligned and the stored `Amount`s were in a specific range and order.

But I'm also not convinced it's worth trying to fix... migration idempotency seems better to solve at a higher level than within individual functions modifying data if it's a design goal, as there are likely other similar cases that will come up in migration code.

Reply:
Ah there's some Discord discussion about this already: https://discord.com/channels/824484045370818580/1052302055680790568/1228444060113698938

Right now, migrations are not idempotent. This may require publishing the post-upgrade root hash, however that doesn't account for migrations affecting non-verifiable storage.

Reply:
We should not tell people migrations are idempotent.

The right way to handle idempotency, later, is probably to use the ABCI application version and compare against a version number in the state.